cmd/format-fs.go

```diff
@@ -0,0 +1,235 @@
+	} `json:"fs"`
+}
+
+// Returns the latest "fs" format.
+func newFormatFS() (format *formatFSV1) {
+	f := &formatFSV1{}
+	f.Version = formatMetaVersionV1
+	f.Format = formatBackendFS
+	f.FS.Version = formatFSVersionV1
+	return f
+}
+
+// Load the format.json
+func formatFSLoad(r io.ReadSeeker, data interface{}) error {
+	if _, err := r.Seek(0, io.SeekStart); err != nil {
+		return err
+	}
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return errors2.Trace(err)
+	}
+	if err = json.Unmarshal(b, data); err != nil {
+		return errors2.Trace(err)
+	}
+	return nil
+}
+
+// Save to format.json
+func formatFSSave(f *os.File, data interface{}) error {
+	b, err := json.Marshal(data)
+	if err != nil {
+		return errors2.Trace(err)
+	}
+	if err = f.Truncate(0); err != nil {
+		return errors2.Trace(err)
+	}
+	_, err = f.Write(b)
+	if err != nil {
+		return errors2.Trace(err)
+	}
+	return nil
+}
+
+// Returns the field formatMetaV1.Format i.e the string "fs" which is never likely to change.
+// We do not use this function in XL to get the format as the file is not fcntl-locked on XL.
+func formatMetaGetFormatBackendFS(r io.ReadSeeker) (string, error) {
+	format := &formatMetaV1{}
+	if err := formatFSLoad(r, format); err != nil {
+		return "", err
+	}
+	if format.Version == formatMetaVersionV1 {
+		return format.Format, nil
+	}
+	return "", fmt.Errorf(`format.Version expected: %s, got: %s`, formatMetaVersionV1, format.Version)
+}
+
+// Returns formatFS.FS.Version
+func formatFSGetVersion(r io.ReadSeeker) (string, error) {
+	format := &formatFSVersionDetect{}
+	if err := formatFSLoad(r, format); err != nil {
+		return "", err
+	}
+	return format.FS.Version, nil
+}
+
+// Migrate the "fs" backend.
+// Migration should happen when formatFSV1.FS.Version changes. This version
+// can change when there is a change to the struct formatFSV1.FS or if there
+// is any change in the backend file system tree structure.
+func formatFSMigrate(wlk *lock.LockedFile) error {
+	// Add any migration code here in case we bump format.FS.Version
+
+	// Make sure that the version is what we expect after the migration.
+	version, err := formatFSGetVersion(wlk)
+	if err != nil {
+		return err
+	}
+	if version != formatFSVersionV1 {
+		return fmt.Errorf(`%s file: expected FS version: %s, found FS version: %s`, formatConfigFile, formatFSVersionV1, version)
+	}
+	return nil
+}
+
+// Creates a new format.json if unformatted.
+func createFormatFS(fsFormatPath string) error {
+	// Attempt a write lock on formatConfigFile `format.json`
+	// file stored in minioMetaBucket(.minio.sys) directory.
+	lk, err := lock.TryLockedOpenFile(fsFormatPath, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return errors2.Trace(err)
+	}
+	// Close the locked file upon return.
+	defer lk.Close()
+
+	fi, err := lk.Stat()
+	if err != nil {
+		return errors2.Trace(err)
+	}
+	if fi.Size() != 0 {
+		// format.json already got created because of another minio process's createFormatFS()
```

Comment:
if you already got lk then there is no way you would see this situation.. since a successful create of format.json also means that you have an active ReadLock.. So you won't get a write lock. I am not sure why you need this code.

Reply:
There is a corner case when two minio processes are calling createFormatFS() in parallel. One process creates the file and releases the lock. At this point the other process should not create this file again.

Reply:
When it is zero bytes why do we care ? Can you provide a reproducer for this.

Reply:
If the lock is already given.  The lk should be preserved and returned up IMO. We shouldnt close it in this function itself ..

Reply:
In createFormatFS() we hold write lock, but in initFormatFS() we need to return read lock. Hence we can't return the write lock held in createFormatFS(). This logic is same as the previous code implemented by you.

Reply:
> When it is zero bytes why do we care ? Can you provide a reproducer for this.

Did not get this. 

```
// Creates a new format.json if unformatted.
func createFormatFS(fsFormatPath string) error {
        // Attempt a write lock on formatConfigFile `format.json`
        // file stored in minioMetaBucket(.minio.sys) directory.
        lk, err := lock.TryLockedOpenFile(fsFormatPath, os.O_RDWR|os.O_CREATE, 0600)
        if err != nil {
                return errors2.Trace(err)
        }
        // Close the locked file upon return.
        defer lk.Close()

        fi, err := lk.Stat()
        if err != nil {
                return errors2.Trace(err)
        }
        if fi.Size() != 0 {
```

At this point there are two possibilities:
1) A competing createFormatFS() has created format.json and written contents to it too. Hence we can just return here.
2) A competing createFormatFS() has created format.json but before it could hold the lock, we hold the lock in which case the size is 0 and we have to write the contents of format.json

This logic has not changed from the previous code.


```
                // format.json already got created because of another minio process's createFormatFS()
                return nil
        }

        return formatFSSave(lk.File, newFormatFS())
}
```

Reply:
Got it..